test/e2e/ics_send.go

```diff
@@ -22,63 +22,56 @@ import (
 	"sigs.k8s.io/yaml"
 )
 
-func sendEventFeature(featureName string, opts sendEventOptions) *feature.Feature {
-	f := feature.NewFeatureNamed(featureName)
+// SendEventFeature will create a feature.Feature that will test sending an
+// event using in cluster sender to SystemUnderTest.
+func SendEventFeature(sut SystemUnderTest) *feature.Feature {
+	f := feature.NewFeatureNamed(sut.Name())
 	sinkName := feature.MakeRandomK8sName("sink")
 	ev := cetest.FullEvent()
 	ev.SetID(feature.MakeRandomK8sName("test-event"))
 
-	f.Setup("deploy sink", eventshub.Install(sinkName, eventshub.StartReceiver))
+	f.Setup("Deploy EventsHub Sink", eventshub.Install(sinkName, eventshub.StartReceiver))
+
+	sink := sut.Deploy(f, sinkName)
 
 	f.Alpha("Event").
-		Must("send", sendEvent(ev, opts.sink(sinkName))).
+		Must("send", sendEvent(ev, sink)).
 		Must("receive", receiveEvent(ev, sinkName))
 
-	for _, st := range opts.steps {
-		f = st(f, sinkName)
-	}
-
 	return f
 }
 
-type (
-	sinkProducer func(string) string
-	step         func(*feature.Feature, string) *feature.Feature
-)
-
-type sendEventOptions struct {
-	sink  sinkProducer
-	steps []step
-}
-
-func sinkFormat(format string) sinkProducer {
-	return func(sinkName string) string {
-		return fmt.Sprintf(format, sinkName)
-	}
-}
-
-func sendEvent(ev cloudevents.Event, sink string) feature.StepFn {
+func sendEvent(ev cloudevents.Event, sink Sink) feature.StepFn {
 	return func(ctx context.Context, t feature.T) {
-		log := logging.FromContext(ctx)
+		log := logging.FromContext(ctx).
+			With(json("event-id", ev.ID()))
+
 		ns := environment.FromContext(ctx).Namespace()
 		args := []string{
 			"send",
 			"--id", ev.ID(),
 			"--source", ev.Source(),
 			"--type", ev.Type(),
 			"--namespace", ns,
-			"--sender-namespace", ns,
 			"--field", fmt.Sprintf("data=%s", ev.Data()),
-			"--to", sink,
+			"--to", sink.String(),
```

Comment:
I don't get why `sink` is a function that can produce string instead of being a simple string from the beginning. Do you need some late binding? From the code it looks like everything should be ready from the beginning.

Reply:
OK. Thanks for the details. Anyways, the setup of the tests is still the same. First, the SUT is deployed and only when it's ready the "sink" is passed to `sendEvent`:
```
sink := sut.Deploy(f, sinkName)

f.Alpha("Event").
	Must("send", sendEvent(ev, sink)).
	Must("receive", receiveEvent(ev, sinkName))
```
So, even if we use GenerateName or something else we will know the actual sink before passing it to sendEvent. Right now I can't think of a use case where the name would change in the meantime (it would be probably unwanted). Let's just simplify it and make the code easier to understand. Thanks

Reply:
Actually, no. It's because of how reconciler-test framework works. The reconciler-test framework has two main phases: Test definition, and test execution.

The following is just test definition. We are only defining steps to live under specific reconciler-test structure:

```golang
sink := sut.Deploy(f, sinkName)

f.Alpha("Event").
	Must("send", sendEvent(ev, sink)).
	Must("receive", receiveEvent(ev, sinkName))
```

At this point, there is no connection to the cluster. This phase produces `*feature.Feature` objects.

When those features gets executed by calling `env.Test` method, test start the execution, and steps functions are getting called:

```golang
env.Test(ctx, t, e2e.SendEventToBroker())
// after avove call we jump inside e2e.SendEventToBroker:
func (b brokerSutImpl) step(ctx context.Context, t feature.T) {
	b.deployBroker(ctx, t)
	b.deployTrigger(ctx, t)
}
```

Current structure allows us to save some information in those private structs (like `brokerSutImpl`) to be used later when `Sink` function is resolved.

Reply:
Hmm. That raises a question: Should `sink := sut.Deploy(f, sinkName)` be called/wrapped within a `f.Setup("Deploy SUT", ...)` ? Right now it's being deployed as part of defining the Feature, but it should probably be done within Setup phase? It's part of the system that needs to be deployed ("set up") before running those tests.

Reply:
Like for example this test in Eventing: https://github.com/knative/eventing/blob/main/test/rekt/features/broker/source_to_sink.go#L41 
The way you did it works as well. I'm not an expert on this REKT framework but I think the intention is to do "setup" through `f.Setup()`

Reply:
> Hmm. That raises a question: Should `sink := sut.Deploy(f, sinkName)` be called/wrapped within a `f.Setup("Deploy SUT", ...)`?

No. it has that code within itself:

```golang
func (bs brokerSut) Deploy(f *feature.Feature, sinkName string) Sink {
  b := brokerSutImpl{sinkName}
  f.Setup("Deploy Broker", b.step)
  return b.sink()
}
```